modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -1141,6 +1141,89 @@ def _apply_func_to_list_of_partitions(cls, func, partitions, **kwargs):
 
+    @classmethod
+    def combine(cls, partitions):
+        """
+        Convert a NumPy 2D array of partitions to a NumPy 2D array of a single partition.
+
+        Parameters
+        ----------
+        partitions : np.ndarray
+            The partitions which have to be converted to a single partition.
+
+        Returns
+        -------
+        np.ndarray
+            A NumPy 2D array of a single partition.
+        """
+
+        def to_pandas_remote(data, partition_shape, *partition_data):
+            """Copy of ``cls.to_pandas()`` method adapted for a remote function."""
+            if all(
+                isinstance(obj, (pandas.DataFrame, pandas.Series))
+                for obj in partition_data
+            ):
+                height, width, *_ = tuple(partition_shape) + (0,)
+                # restore 2d array
+                objs = iter(partition_data)
+                partition_data = [
+                    [next(objs) for _ in range(width)] for __ in range(height)
+                ]
+            else:
+                # Partitions do not always contain pandas objects, for example, hdk uses pyarrow tables.
+                # This implementation comes from the fact that calling `partition.get`
+                # function is not always equivalent to `partition.to_pandas`.
+                partition_data = [
+                    [obj.to_pandas() for obj in part] for part in partition_data
+                ]
+            if all(
+                isinstance(part, pandas.Series)
+                for row in partition_data
+                for part in row
+            ):
+                axis = 0
+            elif all(
+                isinstance(part, pandas.DataFrame)
+                for row in partition_data
+                for part in row
+            ):
+                axis = 1
+            else:
+                ErrorMessage.catch_bugs_and_request_email(True)
+
+            def is_part_empty(part):
+                return part.empty and (
+                    not isinstance(part, pandas.DataFrame) or (len(part.columns) == 0)
+                )
+
+            df_rows = [
+                pandas.concat([part for part in row], axis=axis, copy=False)
+                for row in partition_data
+                if not all(is_part_empty(part) for part in row)
+            ]
+            if len(df_rows) == 0:
+                return pandas.DataFrame()
+            else:
+                return concatenate(df_rows)
+
+        preprocessed_func = cls.preprocess_func(to_pandas_remote)
+        partition_shape = partitions.shape
+        partitions_flattened = partitions.flatten()
+        for idx, part in enumerate(partitions_flattened):
+            if hasattr(part, "force_materialization"):
+                partitions_flattened[idx] = part.force_materialization()
+        partition_refs = [
+            partition.list_of_blocks[0] for partition in partitions_flattened
+        ]
```

Comment:
Why not use `finalize` for this?

Reply:
Removed finalize because when calling partition.list_of_blocks the drain_call_queue() is called in  list_of_blocks function call.

Reply:
> Removed finalize because when calling partition.list_of_blocks the drain_call_queue() is called in list_of_blocks function call.

Ok.

`force_materialization` is also called in `def list_of_blocks` (through `list_of_block_partitions`). Can we remove `force_materialization` call here?

Reply:
In the case when axis partitions are composed of other axis partitions as [here ](https://github.com/arunjose696/modin/blob/master/modin/core/dataframe/pandas/partitioning/axis_partition.py#L119), the partitions would not be force_materialized, and would have multiple partitions as the current implementation of converting to_pandas   requires partitions to have only single partition(https://github.com/modin-project/modin/pull/6880/files#diff-ff5fbf152c79b8591d2ff791f04cda1a8db8e1f082271c98373d6b7f863fd11cR1216) this would not work. 

Also I could see a comment about removing force_materialization() might increase the complexity 
https://github.com/modin-project/modin/pull/4776#discussion_r1032628169

Reply:
got it, thanks! ok for now.